src/java/com/palantir/cassandra/utils/OwnershipVerificationUtils.java

```diff
@@ -37,72 +39,75 @@
 import org.apache.cassandra.utils.FBUtilities;
 import org.apache.cassandra.utils.Hex;
 
-public class MutationVerificationUtils
+public class OwnershipVerificationUtils
 {
     private static final boolean VERIFY_KEYS_ON_WRITE = Boolean.getBoolean("palantir_cassandra.verify_keys_on_write");
-    private static final Logger logger = LoggerFactory.getLogger(MutationVerificationUtils.class);
+    private static final boolean VERIFY_KEYS_ON_READ = Boolean.getBoolean("palantir_cassandra.verify_keys_on_read");
+    private static final Logger logger = LoggerFactory.getLogger(OwnershipVerificationUtils.class);
 
     private static volatile Instant lastTokenRingCacheUpdate = Instant.MIN;
 
-    private MutationVerificationUtils()
+    private OwnershipVerificationUtils()
     {
     }
 
+    public static void verifyRead(Keyspace keyspace, ByteBuffer key)
+    {
+        if (!VERIFY_KEYS_ON_READ)
+        {
+            return;
+        }
+        verifyOperation(keyspace, key, ReadVerificationHandler.INSTANCE);
+    }
+
     public static void verifyMutation(Mutation mutation)
     {
         if (!VERIFY_KEYS_ON_WRITE)
         {
             return;
         }
+        verifyOperation(Keyspace.open(mutation.getKeyspaceName()), mutation.key(), MutationVerificationHandler.INSTANCE);
+    }
 
-        Keyspace keyspace = Keyspace.open(mutation.getKeyspaceName());
+    private static void verifyOperation(Keyspace keyspace, ByteBuffer key, OwnershipVerificationHandler handler)
+    {
         if (!(keyspace.getReplicationStrategy() instanceof NetworkTopologyStrategy))
         {
             return;
         }
 
-        Token tk = StorageService.getPartitioner().getToken(mutation.key());
-        List<InetAddress> cachedNaturalEndpoints = StorageService.instance.getNaturalEndpoints(mutation.getKeyspaceName(), tk);
-        Collection<InetAddress> pendingEndpoints = StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, mutation.getKeyspaceName());
+        String keyspaceName = keyspace.getName();
+        Token tk = StorageService.getPartitioner().getToken(key);
+        List<InetAddress> cachedNaturalEndpoints = StorageService.instance.getNaturalEndpoints(keyspaceName, tk);
+        Collection<InetAddress> pendingEndpoints = StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, keyspaceName);
 
-        if (mutationIsInvalid(cachedNaturalEndpoints, pendingEndpoints))
+        if (operationIsInvalid(cachedNaturalEndpoints, pendingEndpoints))
         {
             if (cacheWasRecentlyRefreshed())
             {
-                throwInvalidMutationException(mutation, keyspace, cachedNaturalEndpoints, pendingEndpoints);
+                handler.onViolation(keyspace, key, cachedNaturalEndpoints, pendingEndpoints);
```

Comment:
an invalid read won't throw an exception here so it will constantly trigger `refreshCache` on every read, which could potentially be problematic. similarly it will still call `handler.onValid(keyspace);` at the botton

i think we should define a `InvalidReadException` and handle it here and below without a refresh

Reply:
good catch! To avoid depending on the presence of an exception for correct control flow, I just added some returns in the correct place.